### PR TITLE
Some refactorings. No functional changes.

### PR DESCRIPTION
**Pull request by bazhenov**
Some cleanups and clippy lints fixed

- some imports reorganized
- unneeded trait bounds, lifetimes, and transformations over iterators removed
- several explanations added to the code
- removed `black_box()`. Returned from the bench function value is [automatically wrapped](https://github.com/bheisler/criterion.rs/blob/e1a8c9ab2104fbf2d15f700d0038b2675054a2c8/src/bencher.rs#L88) by criterion in `black_box()`

### CODE REVIEW

File: benches/search_comparison.rs [hunk 1]
```diff
@@ -167,111 +165,99 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
+    T: TryFrom<usize> + Ord + Copy,
 {
-    group.bench_with_input(BenchmarkId::new(name, size), size, |b, &size| {
+    group.bench_with_input(BenchmarkId::new(name, size), &size, |b, &size| {
         // increasing sequence of even numbers, bounded by MAX
         let iter = (0usize..)
             // Generate only even numbers to provide a ~50% hit ratio in the benchmark
             .map(|i| (i * 2) % MAX)
-            .map(|i| T::try_from(i).unwrap());
+            .map(|i| T::try_from(i).ok().unwrap());
```
Comment (jonhoo) [inline benches/search_comparison.rs:178]: Why this change?

File: src/lib.rs [hunk 2]
```diff
@@ -112,13 +112,12 @@
...
-#[cfg(feature = "nightly")]
-extern crate test;
 
+extern crate alloc;
 #[cfg(test)]
 extern crate std;
-
-extern crate alloc;
+#[cfg(feature = "nightly")]
+extern crate test;
```
Comment (jonhoo) [inline src/lib.rs]: Do we even need this any more after the move to criterion?

File: benches/search_comparison.rs [hunk 3]
```diff
@@ -167,111 +165,99 @@ fn search_bench_case<const MAX: usize, T, Coll>(
...
-        let mut r = pseudorandom_iter::<T>(std::cmp::min(size * 2, MAX));
+        // to generate ~50% hit ratio generated number should be in the range 0..(size * 2) because
+        // test payload contains only even numbers: [0, 2, ..., 2 * size]
+        let mut r = pseudorandom_iter::<T>(MAX.min(size * 2));
         let c = setup_fun(v);
-        b.iter(|| {
-            let x = r.next().unwrap();
-            let _res = black_box(search_fun(&c, x));
-        })
+        b.iter(|| search_fun(&c, r.next().unwrap()))
```
Comment (jonhoo) [inline benches/search_comparison.rs:192]: Do you think we ought to black box the second argument to seatch_fun here?

**Review by jonhoo — APPROVED** [review]

File: benches/search_comparison.rs (hunk 1 above)
Comment (bazhenov) [reply benches/search_comparison.rs:178]: This change converts `Error<T, E>` into `Option<T>`. This way there is no need in `<T as TryFrom<usize>>::Error: core::fmt::Debug` trait because of `unwrap()`, and functions signatures are a little bit more concise. From diagnostic point of view unwrapping the `Result` is the same as unwrapping `Option`, there is no any additional information in `Result`. At least in this version of code.

File: src/lib.rs (hunk 2 above)
Comment (bazhenov) [reply src/lib.rs]: Nice catch! Fixed.

File: benches/search_comparison.rs (hunk 3 above)
Comment (bazhenov) [reply benches/search_comparison.rs:192]: What makes you think that? I don't see how optimizer can do any constant folding of `x`. Or am I missing something?

Comment (jonhoo) [reply benches/search_comparison.rs:192]: I just never underestimate the power of future optimizers is all :p

**Review by jonhoo — APPROVED** [review]
